spacy_curated_transformers/models/listeners.py

```diff
@@ -435,75 +528,244 @@ def __init__(
+        # listener will end up using the annotations of the last upstream transformer
+        # pipe that set the  `trf_data` annotation on the Doc object.
+        assert TransformerListener.use_doc_annotations_for_prediction(listener)
+        TransformerListener.set_use_doc_annotations_for_prediction(listener, False)
+
+    @property
+    def frozen_transformer(self) -> bool:
+        return self.layers[0].attrs["_frozen"]
+
+    @frozen_transformer.setter
+    def frozen_transformer(self, value: bool):
+        self.layers[0].attrs["_frozen"] = value
+
+
+def wrapped_transformer_and_listener_init(model: WrappedTransformerAndListener, X, Y):
+    transformer: TransformerModelT = model.layers[0]
+    listener: TransformerListenerModelT = model.layers[1]
+
+    transformer.init(X=X, Y=Y)
+    listener.init(X=X, Y=Y)
+
+
+def wrapped_transformer_and_listener_forward(
+    model: WrappedTransformerAndListener,
+    docs: WrappedTransformerAndListenerInT,
+    is_train: bool,
+) -> Tuple[WrappedTransformerAndListenerOutT, WrappedTransformerAndListenerBackpropT]:
+    transformer: TransformerModelT = model.layers[0]
+    listener: TransformerListenerModelT = model.layers[1]
+    frozen: bool = transformer.attrs["_frozen"]
+    ops = model.ops
+
+    # Follows the same process as `Transformer.update()`.
+    if frozen:
+        outputs = transformer.predict(docs)
+        bp_outputs = None
+        d_outputs = None
+    else:
+        outputs, bp_outputs = transformer(docs, is_train)
+        d_outputs = [
+            [Ragged(ops.alloc_f(t2v.dataXd.shape), t2v.lengths) for t2v in doc_layers]
+            for doc_layers in outputs.all_outputs
+        ]
+
+    def backprop(
+        one_d_outputs: List[List[Ragged]], outputs_to_backprop: Tuple[int, ...]
+    ) -> Any:
+        nonlocal d_outputs
+        nonlocal frozen
+
+        if frozen:
+            return []
+
+        assert bp_outputs is not None
+        assert d_outputs is not None
+        for i in range(len(one_d_outputs)):
+            for j in outputs_to_backprop:
+                d_outputs[i][j].data += one_d_outputs[i][j].data
+
+        d_docs = bp_outputs(d_outputs)
+        return d_docs
+
+    # Set the output directly on the listener so that it can use them for both
+    # training/backprop and prediction.
+    TransformerListener.receive(
+        listener,
+        batch_id=TransformerListener.calculate_batch_id(docs),
+        outputs=outputs,
+        backprop=backprop,
+    )
+    listener_outputs, bp_listener_outputs = listener(docs, is_train)
+    return listener_outputs, bp_listener_outputs
+
+
+def replace_listener_callback(
+    copied_trf_model: TransformerModelT,
+    trf_listener: TransformerListenerModelT,
+    trf_pipe: Any,
+):
+    # To avoid cyclic imports.
+    from ..pipeline.transformer import Transformer
+
+    assert isinstance(trf_pipe, Transformer)
+    assert TransformerListener.is_listener(trf_listener)
+
+    copied_trf_listener = trf_listener.copy()
```

Comment:
Does a normal copy suffice here, given that the listener has internal state that can be modified?

Reply:
The invocation here is `Model.copy()` which internally performs a deep copy of model's attributes, parameters, etc.